onnxruntime/core/graph/graph_viewer.cc

```diff
@@ -10,21 +10,60 @@
 
 #include "core/graph/graph_utils.h"
 
+#include <queue>
+
 namespace onnxruntime {
 
 bool NodeCompare::operator()(const Node* n1, const Node* n2) const {
   return n1->Index() < n2->Index();
 }
 
+struct PriorityNodeCompare {
+  inline bool IsHighPri(const Node* n) const {
+    static const std::unordered_set<std::string> high_pri_ops = {"Shape", "Size"};
+    return high_pri_ops.find(n->OpType()) != high_pri_ops.end();
+  }
+
+  // Used for std::priority_queue
+  // If return false, n1 will be output first
+  // If return true, n2 will be output first
+  bool operator()(const Node* n1, const Node* n2) const {
+    // nodes in global high priorty list will be output first
+    if (IsHighPri(n1) && !IsHighPri(n2)) {
+      return false;
+    }
+    if (!IsHighPri(n1) && IsHighPri(n2)) {
+      return true;
+    }
+
+    // nodes with lower priority value will be output first
+    if (n1->Priority() != n2->Priority()) {
+      return n1->Priority() > n2->Priority();
+    }
+
+    // otherwise, nodes with lower index will be output first
+    return n1->Index() > n2->Index();
+  }
+};
+
 GraphViewer::GraphViewer(const Graph& graph) {
   graph_ = &graph;
   std::vector<const Node*> leaf_nodes;
   for (auto& node : graph_->Nodes()) {
+    // This is a leaf node (without any output node)
     if (node.OutputNodesBegin() == node.OutputNodesEnd()) {
-      // This is a leaf node (without any output node).
       leaf_nodes.push_back(&node);
     }
+    // This is a root node (without any input node)
+    if (node.InputEdgesBegin() == node.InputEdgesEnd()) {
+      root_nodes_.push_back(node.Index());
+    }
   }
+
+  // Reverse the order of input vector, such that forward nodes are visted in ReverseDFS first
+  // This results in an execution order that forward nodes is always ran before the backward nodes
+  std::reverse(leaf_nodes.begin(), leaf_nodes.end());
```

Comment:
In a training graph, there could be multiple leaf nodes. For example, loss function node (representing the end of FW graph), Optimizer nodes (representing the end of BW graphs) . 

In order to enfore the entire FW is exected before BW, we have to ReverseDFS from the loss function nodes first. 

Leaf_nodes is constructed in the order of node_index, the FW leaf nodes ( with smaller index) will be in the front of the list, BW leaf node will be in the tail. Since ReverseDFS is performed from the tail of the list, so we have to reverse the leaf_node list, to ensure FW leaf nodes are vistited first. 

---
In reply to: [471563245](https://github.com/microsoft/onnxruntime/pull/4526#discussion_r471563245) [](ancestors = 471563245)